module/zfs/zio_crypt.c

```diff
@@ -0,0 +1,1987 @@
+ * the MAC is stored in the embedded checksum within the zil_chain_t header. The
+ * salt and IV are generated for the block on bp allocation instead of at
+ * encryption time. In addition, ZIL blocks have some pieces that must be left
+ * in plaintext for claiming while all of the sensitive user data still needs to
+ * be encrypted. The function zio_crypt_init_uios_zil() handles parsing which
+ * which pieces of the block need to be encrypted. All data that is not
+ * encrypted is authenticated using the AAD mechanisms that the supported
+ * encryption modes provide for. In order to preserve the semantics of the ZIL
+ * for encrypted datasets, the ZIL is not protected at the objset level as
+ * described below.
+ *
+ * DNODE ENCRYPTION:
+ * Similarly to ZIL blocks, the core part of each dnode_phys_t needs to be left
+ * in plaintext for scrubbing and claiming, but the bonus buffers might contain
+ * sensitive user data. The function zio_crypt_init_uios_dnode() handles parsing
+ * which which pieces of the block need to be encrypted. For more details about
+ * dnode authentication and encryption, see zio_crypt_init_uios_dnode().
+ *
+ * OBJECT SET AUTHENTICATION:
+ * Up to this point, everything we have encrypted and authenticated has been
+ * at level 0 (or -2 for the ZIL). If we did not do any further work the
+ * on-disk format would be suscptible to attacks that deleted or rearrannged
+ * the order of level 0 blocks. Ideally, the cleanest solution would be to
+ * maintain a tree of authentication MACs going up the bp tree. However, this
+ * presents a problem for raw sends. Send files do not send information about
+ * indirect blocks so there would be no convenient way to transfer the MACs and
+ * they cannot be recalculated on the receive side without the master key which
+ * would defeat the purpose of raw sends in the first place. Instead, for the
+ * indirect levels of the bp tree, we use a regular SHA512 of the MACs from
+ * the level below. We also include some portable fields from blk_prop such as
+ * the lsize and compression algorithm to prevent the data from being
+ * misinterpretted.
+ *
+ * At the objset level, we maintain 2 seperate 256 bit MACs in the
+ * objset_phys_t. The first one is "portable" and is the logical root of the
+ * MAC tree maintianed in the metadnode's bps. The second, is "local" and is
+ * used as the root MAC for the user accounting objects, which are also not
+ * transferred via "zfs send". The portable MAC is sent in the DRR_BEGIN payload
+ * of the send file. The useraccounting code ensures that the useraccounting
+ * info is not present upon a receive, so the local MAC can simply be cleared
+ * out at that time. For more info about objset_phys_t authentication, see
+ * zio_crypt_do_objset_hmacs().
+ *
+ * CONSIDERATIONS FOR DEDUP:
+ * In order for dedup to work, blocks that we want to dedup with one another
+ * need to use the same IV and encryption key, so that they will have the same
```

Comment:
let's say somewhere in this section that blocks in the same clone family will dedup against one another because they use the same master key.